unison-cli/src/Unison/Cli/Monad.hs

```diff
@@ -0,0 +1,349 @@
+-- success or failure.
+--
+-- * It is a resource monad: resources can be acquired with straight-line syntax, and will be released at the end of
+-- do-block.
+--
+-- * It is an IO monad: you can do IO things, but throwing synchronous exceptions is discouraged. Use the built-in
+-- short-circuiting mechanism instead.
+newtype Cli r a = Cli {unCli :: Env -> (a -> IO (ReturnType r)) -> IO (ReturnType r)}
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadIO,
+      MonadReader Env
+    )
+    via ReaderT Env (ContT (ReturnType r) IO)
+
+-- | What a Cli action returns: a value, an instruction to continue processing input, or an instruction to stop
+-- processing input.
+data ReturnType a
+  = Success a
+  | Continue
+  | HaltRepl
+
+-- | The command-line app monad environment.
+--
+-- Get the environment with 'ask'.
+data Env = Env
+  { authHTTPClient :: AuthenticatedHttpClient,
+    codebase :: Codebase IO Symbol Ann,
+    config :: Configurator.Config,
+    credentialManager :: CredentialManager,
+    -- | Generate a unique name.
+    generateUniqueName :: IO Parser.UniqueName,
+    -- | How to load source code.
+    loadSource :: Text -> IO LoadSourceResult,
+    loopStateRef :: IORef LoopState,
+    -- | What to do with output for the user.
+    notify :: Output -> IO (),
+    -- | What to do with numbered output for the user.
+    notifyNumbered :: NumberedOutput -> IO NumberedArgs,
+    runtime :: Runtime Symbol,
+    sandboxedRuntime :: Runtime Symbol,
+    serverBaseUrl :: Maybe Server.BaseUrl,
+    ucmVersion :: UCMVersion
+  }
+  deriving stock (Generic)
+
+-- | The command-line app monad mutable state.
+--
+-- Get the state with 'getLoopState', put it with 'putLoopState', or modify it with 'modifyLoopState'.
+--
+-- There's an additional pseudo @"currentPath"@ field lens, for convenience.
+data LoopState = LoopState
+  { root :: Branch IO,
+    lastSavedRoot :: Branch IO,
+    -- the current position in the namespace
+    currentPathStack :: List.NonEmpty Path.Absolute,
+    -- TBD
+    -- , _activeEdits :: Set Branch.EditGuid
+
+    -- The file name last modified, and whether to skip the next file
+    -- change event for that path (we skip file changes if the file has
+    -- just been modified programmatically)
+    latestFile :: Maybe (FilePath, Bool),
+    latestTypecheckedFile :: Maybe (UF.TypecheckedUnisonFile Symbol Ann),
+    -- The previous user input. Used to request confirmation of
+    -- questionable user commands.
+    lastInput :: Maybe Input,
+    -- A 1-indexed list of strings that can be referenced by index at the
+    -- CLI prompt.  e.g. Given ["Foo.bat", "Foo.cat"],
+    -- `rename 2 Foo.foo` will rename `Foo.cat` to `Foo.foo`.
+    numberedArgs :: NumberedArgs
+  }
+  deriving stock (Generic)
+
+instance
+  {-# OVERLAPS #-}
+  Functor f =>
+  IsLabel "currentPath" ((Path.Absolute -> f Path.Absolute) -> (LoopState -> f LoopState))
```

Comment:
What's the benefit of an overloaded label here as opposed to a regular ol' lens-as-a-value?

Is there some additional benefit? Otherwise I'd personally prefer to keep things simple and just use a regular lens 😄